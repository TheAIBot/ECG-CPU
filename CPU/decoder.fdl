dp decoder(in op : ns(3);
   	   out pcJMP : ns(1);
	   out ALUInstruction  : ns(2);
	   out useIntermediate : ns(1);
	   out storeALU : ns(1);
	   out doBus : ns(1);
	   out RW : ns(1))
{
	sig x : ns(7);
    lookup q : ns(7) = 
	{
		0b0000001, //ADD
		0b0001101, //AND
		0b0000101, //SUB
		0b0000011, //ADDI
		0b0001011, //SRI
		0b1100000, //STORE
		0b1000001, //LOAD
		0b0010000  //JMP
	};

	always{
        x = q(op);

		doBus = x[6];
		RW = x[5]; //Read = 0, Write = 1
		pcJMP = x[4];
		ALUInstruction = x[2:3];
		useIntermediate = x[1];
		storeALU = x[0];
		//$display($dec, doBus, RW);
	}
}

dp testDecoder(out op : ns(3);
   	       in pcJMP : ns(1);
	       in ALUInstruction  : ns(2);
	       in useIntermediate : ns(1);
	       in storeALU : ns(1)) {
	always{$display($bin, "Cycle:", $cycle, ". Testing decoder for the CPU. op = ", op , ", pcJMP = ", pcJMP , " ALUInstruction = ", ALUInstruction, ", useIntermediate = ", useIntermediate, ", storeALU = ", storeALU);}
	sfg test_ADD	      		{ op = 0b000;}
	sfg test_AND 			{ op = 0b001;}
	sfg test_SUB 			{ op = 0b010;}
	sfg test_ADDI			{ op = 0b011;}
	sfg test_SHIFT_RIGHT_I 		{ op = 0b100;}
	sfg test_LOAD			{ op = 0b101;}
	sfg test_STORE			{ op = 0b110;}
	sfg test_JMP			{ op = 0b111;}
}

// state machine to control testbench
fsm f_testbench(testDecoder){
	initial sADD;
	state sAND, sSUB, sADDI, sSHIFT_RIGHT_I, sLOAD, sSTORE, sJMP;
	@sADD 	    	 (test_ADD)		 -> sAND;
	@sAND 		 (test_AND) 		 -> sSUB;
	@sSUB 		 (test_SUB) 		 -> sADDI;
	@sADDI 		 (test_ADDI) 		 -> sSHIFT_RIGHT_I;
	@sSHIFT_RIGHT_I	 (test_SHIFT_RIGHT_I) 	 -> sLOAD;
	@sLOAD 		 (test_LOAD) 		 -> sSTORE;
	@sSTORE	 	 (test_STORE) 		 -> sJMP;
	@sJMP 		 (test_JMP) 		 -> sADD;
}

//system myFirstSystem {
//	decoder(op, pcJMP, ALUInstruction, useIntermediate, storeALU);
//	testDecoder(op, pcJMP, ALUInstruction, useIntermediate, storeALU);
//}
