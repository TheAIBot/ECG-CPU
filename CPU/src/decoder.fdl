dp decoder(in op : ns(3);
   	   out pcJMP : ns(1);
	   out ALUInstruction  : ns(2);
	   out useIntermediate : ns(1);
	   out storeALU : ns(1);
	   out doBus : ns(1);
	   out RW : ns(1))
{
	sig x : ns(7);
    lookup q : ns(7) = 
	{
		0b0000001, //ADD
		0b0001101, //AND
		0b0000101, //SUB
		0b0000011, //ADDI
		0b0001011, //SRI
		0b1100000, //STORE
		0b1000001, //LOAD
		0b0010000  //JMP
	};

	always{
        x = q(op);

		doBus = x[6];
		RW = x[5]; //Read = 0, Write = 1
		pcJMP = x[4];
		ALUInstruction = x[2:3];
		useIntermediate = x[1];
		storeALU = x[0];
		//$display($dec, doBus, RW);
	}
}
