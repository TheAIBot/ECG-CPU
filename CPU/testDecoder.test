#!/bin/bash
name="testDecoder"
cpp -P $name".fdl" | fdlsim 8 > result.txt
java -jar testVerifier.jar $name".expected" result.txt
if [ $? = "0" ]; then
    echo "Success $name"
else
	echo "Failed $name"
fi
