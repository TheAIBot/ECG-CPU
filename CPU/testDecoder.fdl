#include "src/decoder.fdl"

dp testDecoder(out op : ns(3);
   	       in pcJMP : ns(1);
	       in ALUInstruction  : ns(2);
	       in useIntermediate : ns(1);
	       in storeALU : ns(1);
		   in useMem : ns(1);
		   in RW : ns(1)) {
	always{$display($bin, "Cycle:", $cycle, ". Testing decoder for the CPU. op = ", op , ", pcJMP = ", pcJMP , " ALUInstruction = ", ALUInstruction, ", useIntermediate = ", useIntermediate, ", storeALU = ", storeALU, " useMem = ", useMem, " RW = ", RW);}
	sfg test_ADD	      	{ op = 0b000;}
	sfg test_AND 			{ op = 0b001;}
	sfg test_SUB 			{ op = 0b010;}
	sfg test_ADDI			{ op = 0b011;}
	sfg test_SHIFT_RIGHT_I 	{ op = 0b100;}
	sfg test_LOAD			{ op = 0b101;}
	sfg test_STORE			{ op = 0b110;}
	sfg test_JMP			{ op = 0b111;}
}

// state machine to control testbench
fsm f_testbench(testDecoder){
	initial sADD;
	state sAND, sSUB, sADDI, sSHIFT_RIGHT_I, sLOAD, sSTORE, sJMP;
	@sADD 	    	 (test_ADD)		 -> sAND;
	@sAND 		 (test_AND) 		 -> sSUB;
	@sSUB 		 (test_SUB) 		 -> sADDI;
	@sADDI 		 (test_ADDI) 		 -> sSHIFT_RIGHT_I;
	@sSHIFT_RIGHT_I	 (test_SHIFT_RIGHT_I) 	 -> sLOAD;
	@sLOAD 		 (test_LOAD) 		 -> sSTORE;
	@sSTORE	 	 (test_STORE) 		 -> sJMP;
	@sJMP 		 (test_JMP) 		 -> sADD;
}

system myFirstSystem {
	decoder(op, pcJMP, ALUInstruction, useIntermediate, storeALU, useMem, RW);
	testDecoder(op, pcJMP, ALUInstruction, useIntermediate, storeALU, useMem, RW);
}
