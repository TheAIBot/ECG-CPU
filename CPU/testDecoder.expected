Cycle:0. Testing decoder for the CPU. op = 000, pcJMP = 0 ALUInstruction = 00, useIntermediate = 0, storeALU = 1 useMem = 0 RW = 0
Cycle:1. Testing decoder for the CPU. op = 001, pcJMP = 0 ALUInstruction = 11, useIntermediate = 0, storeALU = 1 useMem = 0 RW = 0
Cycle:2. Testing decoder for the CPU. op = 010, pcJMP = 0 ALUInstruction = 01, useIntermediate = 0, storeALU = 1 useMem = 0 RW = 0
Cycle:3. Testing decoder for the CPU. op = 011, pcJMP = 0 ALUInstruction = 00, useIntermediate = 1, storeALU = 1 useMem = 0 RW = 0
Cycle:4. Testing decoder for the CPU. op = 100, pcJMP = 0 ALUInstruction = 10, useIntermediate = 1, storeALU = 1 useMem = 0 RW = 0
Cycle:5. Testing decoder for the CPU. op = 101, pcJMP = 0 ALUInstruction = 00, useIntermediate = 0, storeALU = 0 useMem = 1 RW = 1
Cycle:6. Testing decoder for the CPU. op = 110, pcJMP = 0 ALUInstruction = 00, useIntermediate = 0, storeALU = 1 useMem = 1 RW = 0
Cycle:7. Testing decoder for the CPU. op = 111, pcJMP = 1 ALUInstruction = 00, useIntermediate = 0, storeALU = 0 useMem = 0 RW = 0
